### ISSUE TITLE
Cleanup: Place::is_indirect vs Place::has_deref

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#114401]
Since https://github.com/rust-lang/rust/pull/97077 we have these two methods `is_indirect` and `has_deref` that are documented as doing the exact same thing, but don't even mention each other in the documentation. Calling the wrong one can either lead to slow code or to subtle bugs. That's pretty bad, we should find a way to clean this up -- and at the very least, this needs better documentation.

**Comment by saethlin** [comment]
I _think_ this is documented in `Place::has_deref`: https://github.com/rust-lang/rust/blob/474709a9a2a74a8bcf0055fadb335d0ca0d2d939/compiler/rustc_middle/src/mir/mod.rs#L1595-L1603

Though why we have that assertion in only one place and not all the others, I have no idea.

**Comment by saethlin** [comment]
I tried pasting this check into `is_indirect` and it fails in `CopyProp` so yeah I have no idea :shrug:

**Comment by RalfJung** [comment]
That doc comment doesn't really tell a user much, does it? It should start by describing what the function does, which is the same as `is_indirect`. And then it should follow "WARNING: only call this in MIR passes with `MirPhase >= Derefered`! It will produce wrong results afterwards."

The current comment is an implementation comment that could go inside the function, but not suitable as a doc comment.

**Comment by RalfJung** [comment]
Also the methods on `Place` should probably just delegate to the ones on `PlaceRef` to avoid the duplication.

**Comment by RalfJung** [comment]
> I tried pasting this check into `is_indirect` and it fails in `CopyProp` so yeah I have no idea shrug

Uh, but doesn't that run sufficiently late that it should be allowed to use `has_deref`?

**Comment by saethlin** [comment]
> Uh, but doesn't that run sufficiently late that it should be allowed to use `has_deref`?

Yes. At a guess, there is MIR which is temporarily invalid during a pass.

**Comment by RalfJung** [comment]
Wow. That makes `has_deref` quite the footgun.
